SU2_CFD/src/driver_adjoint_singlezone.cpp

```diff
@@ -0,0 +1,521 @@
+
+  }
+
+  if (config->GetKind_Solver() == DISC_ADJ_FEM){
+
+    /*--- Apply the boundary condition to clamped nodes ---*/
+    iteration->Postprocess(output,integration_container,geometry_container,solver_container,numerics_container,config_container,surface_movement,grid_movement,FFDBox,ZONE_0,INST_0);
+
+  }
+
+}
+
+void CDiscAdjSinglezoneDriver::SetRecording(unsigned short kind_recording){
+
+  AD::Reset();
+
+  /*--- Prepare for recording by resetting the solution to the initial converged solution---*/
+
+  iteration->SetRecording(solver_container, geometry_container, config_container, ZONE_0, INST_0, kind_recording);
+
+  /*---Enable recording and register input of the iteration --- */
+
+  if (kind_recording != NONE){
+
+    AD::StartRecording();
+
+    if (rank == MASTER_NODE && kind_recording == MainVariables) {
+      cout << endl << "-------------------------------------------------------------------------" << endl;
+      cout << "Direct iteration to store the primal computational graph." << endl;
+      cout << "Compute residuals to check the convergence of the direct problem." << endl;
+    }
+    iteration->RegisterInput(solver_container, geometry_container, config_container, ZONE_0, INST_0, kind_recording);
+
+  }
+
+  /*--- Set the dependencies of the iteration ---*/
+
+  iteration->SetDependencies(solver_container, geometry_container, numerics_container, config_container, ZONE_0, INST_0, kind_recording);
+
+  /*--- Do one iteration of the direct solver ---*/
+
+  DirectRun(kind_recording);
+
+  // NOTE: The inverse design calls were moved to DirectRun() - postprocess
+
+  /*--- Store the recording state ---*/
+
+  RecordingState = kind_recording;
+
+  /*--- Register Output of the iteration ---*/
+
+  iteration->RegisterOutput(solver_container, geometry_container, config_container, output, ZONE_0, INST_0);
+
+  /*--- Extract the objective function and store it --- */
+
+  SetObjFunction();
+
+  AD::StopRecording();
+
+}
+
+void CDiscAdjSinglezoneDriver::SetAdj_ObjFunction(){
+
+  bool time_stepping = config->GetUnsteady_Simulation() != STEADY;
+  unsigned long IterAvg_Obj = config->GetIter_Avg_Objective();
+  unsigned long ExtIter = config->GetExtIter();
+  su2double seeding = 1.0;
+
+  if (time_stepping){
+    if (ExtIter < IterAvg_Obj){
+      seeding = 1.0/((su2double)IterAvg_Obj);
+    }
+    else{
+      seeding = 0.0;
+    }
+  }
+
+  if (rank == MASTER_NODE){
+    SU2_TYPE::SetDerivative(ObjFunc, SU2_TYPE::GetValue(seeding));
+  } else {
+    SU2_TYPE::SetDerivative(ObjFunc, 0.0);
+  }
+
+}
+
+void CDiscAdjSinglezoneDriver::SetObjFunction(){
+
+  bool compressible = (config->GetKind_Regime() == COMPRESSIBLE);
+  bool heat         = (config->GetWeakly_Coupled_Heat());
+  bool turbo        = (config->GetBoolTurbomachinery());
+
+  ObjFunc = 0.0;
+
+  /*--- Specific scalar objective functions ---*/
+
+  switch (config->GetKind_Solver()) {
+  case EULER:                    case NAVIER_STOKES:                   case RANS:
```

Comment:
You mention moving some items to the new output classes (I have seen this in the IO branch), but is it possible to move all of these switches and conditionals to the output too for holding the obj functions, such that one can simply call a virtual function in the base output class like output->SetObjFunc() here and not worry about the particular solver? We have discussed this before and maybe this is the longer term plan..